test/eventlisteners.js

```diff
@@ -169,4 +169,18 @@ describe('event listeners', function() {
     elm.firstChild.click();
     assert.equal(3, result.length);
   });
+  it('Rewrite listener on the same node', function() {
+    var result = [];
+    function onClick(ev) {  };
+    function onClick2(ev) { result.push(ev) };
+    var obj = {on: {click: onClick }};
```

Comment:
this is the main issue here IMO, you're reusing `data` instead of reusing `on`. I don't think you want `data` to be memoizable because you loose the opportunity to do other optimizations. What if you do: `var on = { click: onClick }` and then everytime you construct `h` you pass `{on}` as data?

Also, I don't see the `result.push(ev)` in `onClick`, should we add it to make sure that we are not adding the same listeners over and over again?